### PR TITLE
Major update

### PR DESCRIPTION
**Pull request by Elliot-D-Hill**
This pull request (hopefully) marks the last major refactor before we release on PyPi.

### CODE REVIEW

File: src/diversity/abundance.py [hunk 1]
```diff
@@ -133,122 +83,62 @@ def normalized_subcommunity_abundance(self):
         pass
 
 
-class Abundance(IAbundance):
-    """Implements IAbundance for fast, but memory-heavy calculations.
+class AbundanceFromArray(Abundance):
+    """Implements Abundance for fast, but memory-heavy calculations.
 
     Caches counts and (normalized) relative meta- and subcommunity
```
Comment (JasperBraun) [inline src/diversity/abundance.py:89]: Note: caching and storing in memory are two different things

File: src/diversity/utilities.py [hunk 2]
```diff
@@ -4,219 +4,39 @@
...
-def __validate_power_mean_args(weights, items, atol, weight_is_nonzero):
+def __validate_power_mean_args(
+    weights: ndarray, items: ndarray, atol: float, weight_is_nonzero: ndarray
+) -> None:
     """Validates arguments for power_mean.
 
     Parameters
     ----------
     weights, items, atol
         Same as for power_mean.
```
Comment (JasperBraun) [inline src/diversity/utilities.py:39]:
```suggestion
        Same as for power_mean.
    weight_is_nonzero
        Boolean array of same shape as weights, indicating those whose
        absolute value meets or exceeds atol.
```

File: src/diversity/__main__.py [hunk 3]
```diff
@@ -36,48 +35,17 @@ def main(args):
...
-
-    community_views = []
-    for view in args.viewpoint:
-        community_views.append(meta.subcommunities_to_dataframe(view))
-        community_views.append(meta.metacommunity_to_dataframe(view))
-
-    community_views = concat(community_views, ignore_index=True)
-    community_views.viewpoint = community_views.viewpoint.map(
-        lambda v: format(v, ".2f")
+    counts = read_csv(args.input_filepath, sep=None, engine="python", dtype=int64)
```
Comment (JasperBraun) [inline src/diversity/__main__.py:38]: How safe is sep=None?

**Review by JasperBraun — COMMENTED** [review]
Code looks much cleaner since shared.py module and its functionalities are removed. I left mostly minor comments. Some more major points:
- with the removal of shared.py, the current code stores an array of the size of counts (pivotted) up to 4 times (counts, and normalized counts, as well as, Zp and normalized Zp). Is this acceptable? I have encountered the usecase of setting p to be the identity matrix in dimension equal to Z, which would be infeasible in many contexts. The only context I encountered it in, was for a small Z, though, so it wouldn't be a problem there.
- correspondence between row-ordering and column-ordering of counts and similarity matrix is assumed, which is fine in general. Is the use-case addressed where we might want to subset subcommunities and therefore automatically also species? If I understand correctly, that would currently require initialization of a new metacommunity for each subset of subcommunities, where zero rows in the counts table would indicate missing species and not cause any issues with zero division etc.. Does that sound right?

File: src/diversity/abundance.py [hunk 4]
```diff
@@ -133,122 +83,62 @@ def normalized_subcommunity_abundance(self):
...
     abundances at the same time.
```
[the remark this answers is not in the document]
Comment (Elliot-D-Hill) [reply src/diversity/abundance.py]: I have removed self.counts. It is now immediately converted to self.subcommunnity_abundance so that it is free to be garbage collected.

File: src/diversity/abundance.py [hunk 5]
```diff
@@ -133,122 +83,62 @@ def normalized_subcommunity_abundance(self):
...
 
-    @cache
-    def normalized_subcommunity_abundance(self):
-        return self.subcommunity_abundance() / self.subcommunity_normalizing_constants()
+    @cached_property
+    def normalized_subcommunity_abundance(self) -> ndarray:
+        return self.subcommunity_abundance / self.subcommunity_normalizing_constants
 
 
-class SharedAbundance(IAbundance):
```
[the remark this answers is not in the document]
Comment (Elliot-D-Hill) [reply src/diversity/abundance.py:177]: I have added a class `AbundanceFromSparseArray` which allows for the abundance arrays to be a scipy sparse arrays. This covers the case when the abundance is the identity matrix and other very sparse matrices (e.g. bag of words matrix). Now P and P_bar are sparse. Unfortunately, ZP and ZP_bar will still always be dense.

We could use the same trick of normalizing and un-normalizing while overwriting the same memory block, but this feel like unintuitive behavior (i.e., breaks the principle of least astonishment). For example, this could lead to more computations if the user swapped between normalized and un-normalized measures repeatedly in a loop. Worst of all, they wouldn't be aware that this is happening. Ultimately, it's a trade-off between memory or computation/code clarity.

I think that the code already covers many more big-data applications than the R and Julia package could, and with the addition of sparse array this covers an even greater set of use cases, so I am ok with taking the memory hit.

As a last note, `AbundanceFromSpareArray` requires that scipy's sparse array interface is used and not the sparse matrix interface. This is because the [scipy docs](https://docs.scipy.org/doc/scipy/reference/sparse.html) recommend all future work transition to the array interface. Maybe this is worth noting in our README/docs.

File: src/diversity/components.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    @abstractmethod
+    def get_numerator(measure: str):
+        pass
+
+    @abstractmethod
+    def get_denominator(measure: str):
+        pass
+
+
+class FrequencySensitiveComponents(Components):
```
[the remark this answers is not in the document]
Comment (Elliot-D-Hill) [reply src/diversity/components.py:15]: I have flip flopped back and forth so many times on this terminology. The problem is that SimilarityInsensitiveComponents also makes no reference to frequency. My reasoning for keeping frequency in the name is because frequency is what people associate with traditional diversity measures. I think we could just be explicit in the docs and readme that Leinster diversity is both similarity- and frequency-sensitive.

File: src/diversity/metacommunity.py [hunk 7]
```diff
@@ -4,456 +4,206 @@
...
+            fits on the hard drive disk. Use a Callable function if the similarity matrix does not fit
+            on the disk.
+        X:
+            An array, where each pair of rows will be passed to 'similarity' if it is Callable.
+        chunk_size:
+            The number of file lines to process at a time when the similarity matrix
+            is read from a file. Larger chunk sizes are faster, but take more memory.
+        """
+        LOGGER.debug(
+            "make_metacommunity(counts=%s, similarity=%s, X=%s, chunk_size=%s",
```
[the remark this answers is not in the document]
Comment (Elliot-D-Hill) [reply src/diversity/metacommunity.py:71]: I'm fine with removing the logging statements if you are.

File: src/diversity/metacommunity.py [hunk 8]
```diff
@@ -4,456 +4,206 @@
...
-            abundance=abundance, subcommunity_ordering=subcommunities
-        )
-    elif isinstance(similarity, ISimilarity) and shared_array_manager is None:
-        metacommunity = SimilaritySensitiveMetacommunity(
-            abundance=abundance,
-            subcommunity_ordering=subcommunities,
-            similarity=similarity,
+    def __init__(
+        self,
+        counts: Union[DataFrame, ndarray],
```
[the remark this answers is not in the document]
Comment (Elliot-D-Hill) [reply src/diversity/metacommunity.py:44]: See comments on AbundanceFromSparseArray.

File: src/diversity/metacommunity.py [hunk 9]
```diff
@@ -4,456 +4,206 @@
...
         )
-    elif isinstance(similarity, ISimilarity) and isinstance(
-        shared_array_manager, SharedArrayManager
-    ):
-        metacommunity = SharedSimilaritySensitiveMetacommunity(
-            abundance=abundance,
-            subcommunity_ordering=subcommunities,
-            similarity=similarity,
-            shared_array_manager=shared_array_manager,
+        self.abundance: Abundance = make_abundance(counts=counts)
```
[the remark this answers is not in the document]
Comment (Elliot-D-Hill) [reply src/diversity/metacommunity.py]: I may have gotten a little too excited using type hints. I have removed them.

File: src/diversity/metacommunity.py [hunk 10]
```diff
@@ -4,456 +4,206 @@
...
-
-    @abstractmethod
-    def __init__(self, abundance, subcommunity_ordering):
-        self.abundance = abundance
-        self.subcommunity_ordering = subcommunity_ordering
-        self.measure_components = None
 
-    @cache
-    def subcommunity_diversity(self, viewpoint, measure):
+    def subcommunity_diversity(self, viewpoint: float, measure: str) -> ndarray:
```
[the remark this answers is not in the document]
Comment (Elliot-D-Hill) [reply src/diversity/metacommunity.py:85]: The issue is that `@cache` is equivalent to `@lru_cache(maxsize=None)` which means that old values are never evicted from the cache, so if someone calls `subcommunity_diveristy` in a loop we risk a memory leak.

An alternative I've considered is using `@lru_cache(maxsize=A_SMALL_NUMBER)`, which should cache the case when `metacommunity_diversity` is called right after `subcommunity diversity` as it is in `to_dataframe`.  What do you think?

File: src/diversity/metacommunity.py [hunk 11]
```diff
@@ -4,456 +4,206 @@
...
         subcommunity_diversity = self.subcommunity_diversity(viewpoint, measure)
         return power_mean(
             1 - viewpoint,
-            self.abundance.subcommunity_normalizing_constants(),
+            self.abundance.subcommunity_normalizing_constants,
             subcommunity_diversity,
         )
 
-    def subcommunities_to_dataframe(self, viewpoint):
-        """Table containing all subcommunity diversity values.
```
[the remark this answers is not in the document]
Comment (Elliot-D-Hill) [reply src/diversity/metacommunity.py:221]: Not sure why I deleted this. Added it back in a new commit.

File: src/diversity/metacommunity.py [hunk 12]
```diff
@@ -4,456 +4,206 @@
...
-            the same as frequent species, and infinity considers only the
-            most frequent species.
+        viewpoint:
+            Affects the contribution of rare species to the diversity measure.
+            When viewpoint = 0 all species (rare or frequent) contribute equally.
+            When viewpoint = infinity, only the single most frequent species contribute.
+
+        Returns
+        -------
+        A pandas.DataFrame containing all subcommunity diversity measures for a given viewpoint
```
[the remark this answers is not in the document]
Comment (Elliot-D-Hill) [reply src/diversity/metacommunity.py]: I have added a line length ruler to my IDE and updated all the docstrings so they fit within 72 characters.

File: src/diversity/parameters.py [hunk 13]
```diff
@@ -46,27 +46,6 @@ def configure_arguments():
...
-    )
-    parser.add_argument(
-        "-c",
-        "--count_column",
-        help="Header of the column containing the counts.",
-        default="count",
-    )
     parser.add_argument(
         "-i",
         "--input_filepath",
```
[the remark this answers is not in the document]
Comment (Elliot-D-Hill) [reply src/diversity/parameters.py:51]: Fixed.

File: src/diversity/similarity.py [hunk 14]
```diff
@@ -195,407 +52,189 @@ def calculate_weighted_similarities(self, relative_abundances, out=None):
...
+    def weighted_similarities(self, relative_abundances: ndarray) -> ndarray:
+        return self.similarity.to_numpy() @ relative_abundances
 
-class SimilarityFromFile(ISimilarity):
-    """Implements ISimilarity by using similarities stored in file.
+
+class SimilarityFromArray(Similarity):
+    """Implements Similarity using similarities stored in a numpy ndarray."""
+
+    def __init__(self, similarity: ndarray) -> None:
```
[the remark this answers is not in the document]
Comment (Elliot-D-Hill) [reply src/diversity/similarity.py]: I like this suggestion, but unfortunately `ArrayLike` doesn't work with `isinstance()` or `type()` so I can't use it to dispatch `Abundance` classes in `make_abundance`. So, even if we could technically pass a list of lists or some other array-like datatype, we don't know which class to pass it to. I don't think it's too much of an ask to require the user to pass an explicit np.array, but if you have suggestions to make `ArrayLike` work, that would be cool too.

File: src/diversity/similarity.py [hunk 15]
```diff
@@ -195,407 +52,189 @@ def calculate_weighted_similarities(self, relative_abundances, out=None):
...
         ]
-        with Pool(self.__num_processors) as pool:
-            pool.starmap(self.__similarity_function, args_list)
-        if out is not None and not isinstance(out, SharedArrayView):
-            out[:] = out_.data
-            return out
-        else:
-            return out_.data
+        weighted_similarity_chunks = get(futures)
+        return concatenate(weighted_similarity_chunks)
```
[the remark this answers is not in the document]
Comment (Elliot-D-Hill) [reply src/diversity/similarity.py:199]: One limitation of the shared memory approach that Ray uses is that the memory store is immutable so we can't write the chunks directly into a shared array. Another trade-off between memory vs simplicity (and speed, since Ray is slightly faster than the multiprocessing standard library).

File: src/diversity/similarity.py [hunk 16]
```diff
@@ -195,407 +52,189 @@ def calculate_weighted_similarities(self, relative_abundances, out=None):
...
-        """
-        self.__species_ordering = self.__get_species_ordering(
-            similarity_matrix, species_subset
-        )
-        self.similarity_matrix = self.__reindex_similarity_matrix(similarity_matrix)
-
-    @property
-    def species_ordering(self):
-        return self.__species_ordering
+def make_similarity(
```
[the remark this answers is not in the document]
Comment (Elliot-D-Hill) [reply src/diversity/similarity.py:202]: Yes, for example, like X and y in sklearn.

File: src/diversity/__main__.py (hunk 3 above)
Comment (Elliot-D-Hill) [reply src/diversity/__main__.py:38]: I believe `pandas.read_csv` samples from the file and uses `csv.Sniffer` from  the standard library. This approach feels safer than relying on the user to give the correct extension. It also allows for other delimiters (e.g., ";").

**Comment by Elliot-D-Hill** [comment]
I removed filtering because the way we had it previously rippled out into filtering logic in every strategy. By letting the user filter the data before creating a `Metacommunity` object, we remove all of that logic (and all of the tests that test filtering). The downside is that, as you mentioned, they have to create a new object for each new subset.

```python
counts_subset = counts[species, subcommunities]
similarity_matrix_subset = similarity_matrix[species, species]
Metacommunity(counts=counts_subset, similarity=similarity_matrix_subset)
```

I have plans to add filtering back in at some point, but only at `Metacommunity.subcommunity_diverisity`. The idea would be to pass views of P, P_bar, ZP, and ZP_bar (after their initialization) to `power_mean`. I think this would minimize the coupling effect that filtering previously had and save the user from having to recompute the components.
